lightum.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "lightum 1"
.TH lightum 1 "2012-05-01" "lightum v2.0" "User commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
lightum \- MacBook automatic light sensor daemon
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
lightum [\-m \fIvalue\fR] [\-n \fIvalue\fR] [\-M \fIvalue\fR] [\-N \fIvalue\fR] [\-p \fIvalue\fR] [\-i \fIvalue\fR] [\-I \fIvalue\fR] [\-w {1,2,3}] [\-x] [\-u] [\-l] [\-s] [\-f] [\-v] [\-d {1,2,3}]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
lightum is a daemon to control the keyboard brightness and monitor backlight on
MacBook based laptops.
.PP
By default it will run in \*(L"auto\*(R" mode, this means it will read data from the
ambient light sensor (located on the iSight camera) and automatically:
.PP
.Vb 2
\&        a) light up or dim the keyboard brightness 
\&        b) light up or dim the the video backlight
.Ve
.PP
It will also dim the keyboard brightness and screen backlight when the computer
is not used for a configurable amount of seconds (5 by default).
.PP
If you choose to run it in \*(L"manual\*(R" mode, it will not query the ambient light
sensor, but instead it will use the brightness and backlight values you set
manually using the function keys.
.PP
When first launched lightum will create a configuration file in the folder
~/.config/lightum/lightum.conf, inside the user's home directory with the
default configuration values. When launched without parameters, lightum will
read the configuration from this config file, but the configuration values can
be overwritten via the command line.
.SH "OPTIONS"
.IX Header "OPTIONS"
Usage:  lightum [\s-1OPTION\s0]...
     \fB\-m\fR \fI4..255\fR : maximum brightness value in auto mode (default=255)
     \fB\-n\fR \fI0..3\fR   : minimum brightness value in auto mode (default=0)
     \fB\-M\fR \fI7..15\fR  : maximum backlight value in auto mode (default=15)
     \fB\-N\fR \fI1..6\fR   : minimum backlight value in auto mode (default=1)
     \fB\-p\fR \fInum\fR    : number of milliseconds between light sensor polls (default=300)
     \fB\-i\fR \fInum\fR    : power off keyboard light on session idle seconds (0 to disable)
     \fB\-I\fR \fInum\fR    : turn down screen backlight on session idle seconds (0 to disable)
     \fB\-w\fR \fInum\fR    : 1 manage brightness, 2 manage backlight, 3 both (default:3)
     \fB\-x\fR        : manual mode (will honor the brightness value set with Fn keys)
     \fB\-l\fR        : fully dim the screen backlight when session is idle
     \fB\-u\fR        : do not ignore brightness changes happening outside lightum
     \fB\-s\fR        : power off keyboard light when screen saver is active
     \fB\-f\fR        : run in foreground (do not daemonize)
     \fB\-v\fR        : verbose mode, useful for debugging with \-f and \-d
     \fB\-d\fR \fInum\fR    : debug mode: 1 brightness, 2 backlight, 3 both
.SH "CONFIGURATION FILE"
.IX Header "CONFIGURATION FILE"
\&\fBGeneral options\fR:
.PP
.Vb 4
\&     # manualmode
\&     #   0 = automatically adjust keyboard brightness based on light sensor
\&     #   1 = or control keyboard brightness manually using Fn+ F5/F6 keys
\&     manualmode=0
\&     
\&     # ignoreuser: only has effect in auto\-mode (when manualmode=0)
\&     #   0 = change maxbrightness value dinamically when user presses Fn+ F5/F6
\&     #   1 = ignore brightness changess happening outside lightum and keep the
\&     #       maxbrightness value from the config file (fixes bug in ubuntu 12.04)
\&     ignoreuser=1
\&     
\&     # workmode
\&     #   1 = only manage keyboard brightness (ignore screen backlight)
\&     #   2 = only manage screen backlight (ignore keyboard brightness)
\&     #   3 = manage both keyboard brightness and screen backlight
\&     workmode=3
\&     
\&     # poll time in milliseconds (used for light sensor and session idle time)
\&     polltime=300
\&
\&     # screensaver
\&     #   1 = turn off keyboard brightness when screensaver is active
\&     #   0 = do not monitor screensaver status
\&     queryscreensaver=0
.Ve
.PP
\&\fBOptions to manage keyboard brightness\fR:
.PP
.Vb 2
\&     # maximum keyboard brightness value (between 4 and 255)
\&     maxbrightness=56
\&     
\&     # minimum keyboard brightness value (between 0 and 3)
\&     minbrightness=2
\&     
\&     # turn off keyboard brightness if computer unused for X seconds (0 to disable)
\&     idleoff=4
.Ve
.PP
\&\fBOptions to manage screen backlight\fR:
.PP
.Vb 2
\&     # maximum screen backlight value (between 4 and 15)
\&     maxbacklight=12
\&     
\&     # minimum screen backlight value (between 0 and 3)
\&     minbacklight=2
\&     
\&     # turn off screen backlight if computer unused for X seconds (0 to disable)
\&     screenidle=30
\&     
\&     # fully dim the backlight when screenidle seconds reached (default 0)
\&     #   1 = when idle, backlight will be set to 1

debian/changelog
lightum (2.3-ubuntu1) precise; urgency=low

  * updated to version 2.3 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Mon, 14 May 2012 21:11:21 +0200

lightum (2.2.1-ubuntu1) precise; urgency=low

  * updated to version 2.2.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sun, 13 May 2012 17:35:57 +0200

lightum (2.2-ubuntu1) precise; urgency=low

  * updated to version 2.2
  * added libc6-dev build time dependancy for math.h

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sun, 13 May 2012 16:50:49 +0200

lightum (2.1.1-ubuntu1) precise; urgency=low

  * updated to version 2.1.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 11 May 2012 20:11:02 +0200

lightum (2.1-ubuntu1) precise; urgency=low

  * updated to version 2.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Wed, 09 May 2012 21:08:30 +0200

lightum (2.0.4-ubuntu1) precise; urgency=low

  * updated to version 2.0.4 (fixes a memory leak)

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 04 May 2012 22:05:20 +0200

lightum (2.0.3-ubuntu1) precise; urgency=low

  * updated to version 2.0.3
  * added debian/watch file

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Thu, 03 May 2012 16:27:42 +0200

lightum (2.0.2-ubuntu1) precise; urgency=low

  * updated to version 2.0.2
  * added libdbus-glib-1-dev as build time dependancy
  * added consolekit as run time dependancy
  * updated program description

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Wed, 02 May 2012 17:46:59 +0200

lightum (2.0.1-ubuntu1) precise; urgency=low

  * updated to version 2.0.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 01 May 2012 11:49:30 +0200

lightum (2.0-ubuntu1) precise; urgency=low

  * updated to version 2.0

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 01 May 2012 01:16:54 +0200

lightum (1.7-ubuntu1) precise; urgency=low

  * updated to version 1.7

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Thu, 26 Apr 2012 22:28:07 +0200

lightum (1.6-ubuntu1) oneiric; urgency=low

  * updated to version 1.6

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Thu, 26 Apr 2012 16:52:54 +0200

lightum (1.5-ubuntu1) oneiric; urgency=low

  * updated to version 1.5

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 13 Dec 2011 15:45:33 +0100

lightum (1.4.1-ubuntu1) oneiric; urgency=low

  * updated to version 1.4.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Mon, 12 Dec 2011 00:35:01 +0100

lightum (1.4-ubuntu1) oneiric; urgency=low

  * updated to version 1.4

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sat, 10 Dec 2011 03:13:28 +0100

lightum (1.3.1-ubuntu1) oneiric; urgency=low

  * updated to version 1.3.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 09 Dec 2011 23:35:02 +0100

lightum (1.3-ubuntu1) oneiric; urgency=low

  * updated to version 1.3

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 09 Dec 2011 21:00:39 +0100

lightum (1.2-ubuntu1) oneiric; urgency=low

  * updated to version 1.2, removed libdbus-glib-1-dev dependancy, added libxss-dev dependancy.

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Wed, 07 Dec 2011 03:27:00 +0100

lightum (1.1-ubuntu4) oneiric; urgency=low

  * fix build time dependancy, hopefully. Package name is libdbus-glib-1-dev, not dbus-glib-1.

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 05:12:17 +0100

lightum (1.1-ubuntu3) oneiric; urgency=low

  * fix build time dependancy, again: dbus-glib-1 is included in Makefile pkgconfig, but not really needed. 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 04:57:04 +0100

lightum (1.1-ubuntu2) oneiric; urgency=low

  * fix build time dependancy on debian control file (dbus/dbus.h) 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 04:38:48 +0100

lightum (1.1-ubuntu1) oneiric; urgency=low

  * updated to version 1.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 04:17:53 +0100

lightum (1.0-ubuntu2) oneiric; urgency=low

  * Fix Makefile for launchpad build 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sun, 04 Dec 2011 12:04:07 +0100

lightum (1.0-ubuntu1) oneiric; urgency=low

  * Initial release

 -- Pau Oliva Fora (pof) <pau@eslack.org> Sun, 04 Dec 2011 05:23:38 +0100 


lightum.c
/*
 *   MacBook automatic light sensor daemon
 *   Copyright 2011 Pau Oliva Fora <pof@eslack.org>
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/stat.h>

#include "lightum.h"

#define VERSION "2.3"

#define MAXLV	15

int reloadconfig=0;

void usage(const char *error) {
	fprintf(stderr, "lightum v%s - (c)2011 Pau Oliva Fora <pof@eslack.org>\n",VERSION);
	fprintf(stderr, "%s",error);
	fprintf(stderr, "Usage:  lightum [OPTION]...\n");
	fprintf(stderr, "     -m 4..255 : maximum brightness value in auto mode (default=255)\n");
	fprintf(stderr, "     -n 0..3   : minimum brightness value in auto mode (default=0)\n");
	fprintf(stderr, "     -M 7..15  : maximum backlight value in auto mode (default=15)\n");
	fprintf(stderr, "     -N 1..6   : minimum backlight value in auto mode (default=1)\n");
	fprintf(stderr, "     -p num    : number of milliseconds between light sensor polls (default=300)\n");
	fprintf(stderr, "     -i num    : power off keyboard light on session idle seconds (0 to disable)\n");
	fprintf(stderr, "     -I num    : turn down screen backlight on session idle seconds (0 to disable)\n");
	fprintf(stderr, "     -w num    : 1 manage brightness, 2 manage backlight, 3 both (default:3)\n");
	fprintf(stderr, "     -x        : manual mode (will honor the brightness value set with Fn keys)\n");
	fprintf(stderr, "     -l        : fully dim the screen backlight when session is idle\n");
	fprintf(stderr, "     -u        : do not ignore brightness changes happening outside lightum\n");
	fprintf(stderr, "     -U        : ignore session information from ConsoleKit\n");
	fprintf(stderr, "     -s        : power off keyboard light when screen saver is active\n");
	fprintf(stderr, "     -f        : run in foreground (do not daemonize)\n");
	fprintf(stderr, "     -v        : verbose mode, useful for debugging with -f and -d\n");
	fprintf(stderr, "     -d num    : debug mode: 1 brightness, 2 backlight, 3 both\n");
	exit(1);
}

void check_config_values(conf_data conf) {
	if (conf.manualmode < 0 || conf.manualmode > 1) usage("ERROR: Wrong value in config variable 'manualmode'\n");
	if (conf.ignoreuser < 0 || conf.ignoreuser > 1) usage("ERROR: Wrong value in config variable 'ignoreuser'\n");
	if (conf.ignoresession < 0 || conf.ignoresession > 1) usage("ERROR: Wrong value in config variable 'ignoresession'\n");
	if (conf.queryscreensaver < 0 || conf.queryscreensaver > 1) usage("ERROR: Wrong value in config variable 'queryscreensaver'\n");
	if (conf.maxbrightness < 4 || conf.maxbrightness > 255) usage("ERROR: Wrong value in config variable 'maxbrightness'\n");
	if (conf.minbrightness < 0 || conf.minbrightness > 3) usage("ERROR: Wrong value in config variable 'minbrightness'\n");
	if (conf.maxbacklight < 7 || conf.maxbacklight > 15) usage("ERROR: Wrong value in config variable 'maxbacklight'\n");
	if (conf.minbacklight < 1 || conf.minbacklight > 6) usage("ERROR: Wrong value in config variable 'minbacklight'\n");
	if (conf.polltime < 1 || conf.polltime > 100000) usage("ERROR: Wrong value in config variable 'polltime'\n");
	if (conf.idleoff < 0 || conf.idleoff > 86400) usage("ERROR: Wrong value in config variable 'idleoff'\n");
	if (conf.screenidle < 0 || conf.screenidle > 86400) usage("ERROR: Wrong value in config variable 'screenidle'\n");
	if (conf.workmode < 1 || conf.workmode > 3) usage("ERROR: Wrong value in config variable 'workmode'\n");
}

int main(int argc, char *argv[]) {

	int screensaver=0, c, brightness_prev=-1, backlight_prev=-1;
	int light=0, brightness=255, backlight=100;
	int foreground=0, verbose=0, debug=0;
	int brightness_restore, backlight_restore, brightness_restoreflag=0, backlight_restoreflag=0;
	int res, dbus_backend=-1, tmp=-1;
	float idletime=0;
	pid_t pid;
	conf_data conf;
	Display *display = NULL;
        DBusGConnection *connection;
        DBusGProxy *proxy_manager;
        DBusGProxy *proxy_session;
	uid_t uid, euid;
	int light_aux=-1, light_avg=-1;
	int lightvalues[MAXLV] = {0};
	int countarray[255] = {0};
	unsigned int i,index=0;

	// make sure we are run as a regular user
	uid = getuid();
	euid = geteuid();
	if (uid == 0 || euid == 0) {
		fprintf(stderr, "lightum must NOT be run as root.\n");
		exit(1);
	}

	// overwrite defaults with config file
	conf = config_parse();

	// overwrite config file with command line arguments
	while ((c = getopt(argc, argv, "hxuUlsvfm:n:M:N:p:I:i:d:w:?")) != EOF) {
		switch(c) {
			case 'h':
				usage("");
				break;
			case 'x':
				conf.manualmode=1;
				break;
			case 'u':
				conf.ignoreuser=0;
				break;
			case 'U':
				conf.ignoresession=1;
				break;
			case 'l':
				conf.fulldim=1;
				break;
			case 's':
				conf.queryscreensaver=1;
				break;
			case 'f':
				foreground=1;
				break;
			case 'v':
				verbose=1;
				break;
			case 'd':
				debug=atoi(optarg);
				break;
			case 'm':
				conf.maxbrightness=atoi(optarg);
				break;
			case 'n':
				conf.minbrightness=atoi(optarg);
				break;
			case 'M':
				conf.maxbacklight=atoi(optarg);
				break;
			case 'N':
				conf.minbacklight=atoi(optarg);
				break;
			case 'p':
				conf.polltime=atoi(optarg);
				break;
			case 'i':
				conf.idleoff=atoi(optarg);
				break;
			case 'I':
				conf.screenidle=atoi(optarg);
				break;
			case 'w':
				conf.workmode=atoi(optarg);
				break;
			default:
				usage("ERROR: Unknown OPTION\n");
				break;
		}
	}

	if (verbose) printf("CONFIG:\n\tmanualmode: %d\n",conf.manualmode);
	if (verbose) printf("\tignoreuser: %d\n",conf.ignoreuser);
	if (verbose) printf("\tignoresession: %d\n",conf.ignoresession);
	if (verbose) printf("\tworkmode: %d\n",conf.workmode);
	if (verbose) printf("\tqueryscreensaver: %d\n",conf.queryscreensaver);
	if (verbose) printf("\tmaxbrightness: %d\n",conf.maxbrightness);
	if (verbose) printf("\tminbrightness: %d\n",conf.minbrightness);
	if (verbose) printf("\tmaxbacklight: %d\n",conf.maxbacklight);
	if (verbose) printf("\tminbacklight: %d\n",conf.minbacklight);
	if (verbose) printf("\tpolltime: %d\n",conf.polltime);
	if (verbose) printf("\tidleoff: %d\n",conf.idleoff);
	if (verbose) printf("\tscreenidle: %d\n",conf.screenidle);
	if (verbose) printf("\tfulldim: %d\n\n",conf.fulldim);

	// make sure all config values are correct
	check_config_values(conf);
	if (debug < 0 || debug > 3) usage("ERROR: Wrong value in config variable 'debug'\n");

	// if debug enabled, force verbose mode too
	if (debug > 0) verbose=1;
	// if verbose enabled, force foreground mode too
	if (verbose) foreground=1;

	if (conf.manualmode) printf("lightum v%s running in manual mode ", VERSION);
	else printf("lightum v%s running in auto mode ", VERSION);
	fflush(stdout);

	if (!foreground) {
		if ((pid = fork()) < 0) exit(1);
		else if (pid != 0) exit(0);
		/* daemon running here */
		setsid();
		res=chdir("/");
		if (res != 0) {
			perror("Could not chdir");
			exit(1);
		}
		umask(0);
		printf("forked into background\n");
	} else printf("\n");

	/* create pid file */
	if (!create_pid_file()) exit(1);

	/* start with current brightness values */
	if (conf.workmode == 1 || conf.workmode == 3) {
		brightness_restore=get_keyboard_brightness_value();
	}

	/* start with current backlight values */
	if (conf.workmode == 2 || conf.workmode == 3) {
		backlight_restore=get_screen_backlight_value();

		// detect dbus backend: 0: gnome, 1: kde
		tmp = dbus_set_screen_backlight_value_gnome(acpi_to_dbus_backlight(backlight_restore));
		if (tmp == -1) {
			tmp = dbus_set_screen_backlight_value_kde(acpi_to_dbus_backlight(backlight_restore));
			if (tmp == -1) {
				fprintf (stderr, "Can't manage screen backlight on this system.\nPlease disable backlight with config option 'workmode='1' or command line switch '-w 1'.\nIf you believe this is an error, open a bug report: https://github.com/poliva/lightum/issues\n");
				exit (1);
			} else {
				dbus_backend=1;	
			}

		} else {
			dbus_backend=0;
		}
	}

	if (conf.idleoff != 0 || conf.screenidle != 0) {
		display = XOpenDisplay(NULL);
		if (display == NULL) {
			fprintf(stderr,"Failed to open display\n");
			exit(1);
		}
	}

	signal_installer();

	if (!conf.ignoresession) {
		connection = get_dbus_connection();
		proxy_manager = get_dbus_proxy_manager(connection);
		proxy_session = get_dbus_proxy_session(connection, proxy_manager);
	}

	// initialize the light values array
	for (i=0;i<256;i++)
		countarray[i]=0;
	if (!conf.manualmode) {
		light=get_light_sensor_value();
		for (i=0; i<MAXLV; i++) 
			lightvalues[i]=light;
		countarray[light]=MAXLV;
	} else {
		for (i=0; i<MAXLV; i++) 
			lightvalues[i]=0;
	}

	while(1) {

		if (reloadconfig) {
			conf = config_parse();
			if (verbose) printf("lightum: SIGUSR1 received, configuration reloaded\n");
			check_config_values(conf);
			reloadconfig=0;
		}

		if (!conf.ignoresession) {
			if (! get_session_active(proxy_session) ) {
				if (verbose) printf("lightum: user session not active, sleeping %d milliseconds.\nIf you believe this is an error, try running lightum with 'ignoresession=1' or '-U' command line switch.\n", conf.polltime);
				usleep(conf.polltime*1000);
				continue;
			}
		}

		if (!conf.manualmode) {
			light=get_light_sensor_value();
			if (verbose) printf("light_sensor: %d ",light);

			// to avoid backlight flickering when the light sensor flaps too fequently
			// between two values, we collect lighting values and use the most common
			// value of the collected values

			if (index == MAXLV) index=0;
			lightvalues[index] = light;

			// get the most repetitive value of lightvalues array
			for (i = 0; i < MAXLV; i++) {
				countarray[lightvalues[i]]++;
			}

			light_avg=-1;
			light_aux=-1;
			for (i = 0; i < 256; i ++) {
				if(countarray[i] > light_aux) {
					light_aux = countarray[i];
					light_avg=i;
				}
				countarray[i]=0;
			}

			light=light_avg;

			if (verbose) printf("light_avg: %d ",light);
			index++;
		}

		if (conf.idleoff != 0 || conf.screenidle != 0) {
			idletime=get_session_idle_time(display);
			if (verbose) printf("idle_time: %f ",idletime);
		}

		if (!conf.manualmode) {

			if (conf.workmode == 1 || conf.workmode == 3) brightness=calculate_keyboard_brightness_value(light, conf.maxbrightness, conf.minbrightness);
			if (conf.workmode == 2 || conf.workmode == 3) backlight=calculate_screen_backlight_value(light, conf.maxbacklight, conf.minbacklight);
			if (verbose) printf("auto mode ");

		} else {

			if (verbose) printf("manual mode ");
			if (conf.workmode == 1 || conf.workmode == 3) {
				if (!screensaver) {
					if (idletime > conf.idleoff) {
						if (brightness_restoreflag==0) {
							brightness_restore=get_keyboard_brightness_value();
							brightness_restoreflag=1;
							if (debug == 1 || debug == 3) printf("brightness_restoreflag(%d) ", brightness_restore);
						}
						brightness=conf.minbrightness;
					} else {
						brightness=brightness_restore;
						brightness_restoreflag=0;
						if (debug == 1 || debug == 3) printf("brightness_restored(%d) ", brightness_restore);
					}
				}
			}

			if (conf.workmode == 2 || conf.workmode == 3) {
				if (!screensaver) {
					if (idletime > conf.screenidle) {
						if (backlight_restoreflag==0) {
							backlight_restore=get_screen_backlight_value();
							backlight_restoreflag=1;
							if (debug == 2 || debug == 3) printf("backlight_restoreflag(%d) ", backlight_restore);
						}
						if (conf.fulldim) backlight=1;
						else backlight=conf.minbacklight;
					} else {
						backlight=backlight_restore;
						backlight_restoreflag=0;
						if (debug == 2 || debug == 3) printf("backlight_restored(%d) ", backlight_restore);
					}
				}
			}

		}

		if (conf.workmode == 1 || conf.workmode == 3) {
			if ((conf.idleoff != 0) && (idletime > conf.idleoff)) {
				brightness=conf.minbrightness;
			}
		}

		if (conf.workmode == 2 || conf.workmode == 3) {
			if ((conf.screenidle != 0) && (idletime > conf.screenidle)) {
				if (conf.fulldim) backlight=1;
				else backlight=conf.minbacklight;
			}
		}

		if (conf.queryscreensaver) {
			screensaver=get_screensaver_active();
			if (verbose) printf("screensaver: %d ",screensaver);
		 	if (screensaver) {
				brightness=0;
				backlight=conf.minbacklight;
			}
		}

		if (conf.workmode == 1 || conf.workmode == 3)
			if (verbose) printf("brightness: %d/%d ",brightness, conf.maxbrightness);

		if (conf.workmode == 2 || conf.workmode == 3)
			if (verbose) printf("backlight: %d/%d ",backlight, conf.maxbacklight);

		// keyboard brightness
		if (conf.workmode == 1 || conf.workmode == 3) {
			if (brightness!=brightness_prev) {
				if (!conf.manualmode) {
					brightness_restore=get_keyboard_brightness_value();
					if (brightness_restore < conf.minbrightness) brightness_restore=conf.minbrightness;
					if (debug == 1 || debug == 3) printf("\ncurrent brightness: %d\n",brightness_restore);
					if ((brightness_restore != brightness_prev) && (brightness_restoreflag)) {
						if (!conf.ignoreuser) {
							/* make sure maxbrightness is never <4 */
							if (brightness_restore < 4) conf.maxbrightness=4;
							else conf.maxbrightness=brightness_restore;
							if (verbose) printf("-> Detected user brightness change, setting maxbrightness to %d\n",conf.maxbrightness);
							brightness_prev=brightness_restore;
						} else {
							if (verbose) printf("-> Ignoring user brightness change, wants to set maxbrightness to %d\n",brightness_restore);
						}
						brightness=calculate_keyboard_brightness_value(light, conf.maxbrightness, conf.minbrightness);
					}
					brightness_restoreflag=1;
				}
				if (debug == 1 || debug == 3) printf ("-> set keyboard brightness: %d -> %d\n",brightness_prev,brightness);
				fading(brightness_prev,brightness);
				usleep(1500);
				brightness=get_keyboard_brightness_value();
				brightness_prev=brightness;
			}
			if (!conf.manualmode) {
				tmp=get_keyboard_brightness_value();
				if (tmp!=brightness) {
					if (verbose) printf("-> Detected user brightness change, current brightness is set to %d\n", tmp);
					if (conf.ignoreuser) {
						if (debug == 1 || debug == 3) printf ("\n*** forcing brightness from %d to %d\n", tmp, brightness);
						fading(tmp,brightness);
					}
				}
			}
		}

		// screen backlight
		if (conf.workmode == 2 || conf.workmode == 3) {
			if (backlight!=backlight_prev) {
				if (!conf.manualmode) {
					backlight_restore=get_screen_backlight_value();
					if (backlight_restore < conf.minbacklight) backlight_restore=conf.minbacklight;
					if (debug == 2 || debug == 3) printf("\ncurrent backlight: %d\n",backlight_restore);
					if ((backlight_restore != backlight_prev) && (backlight_restoreflag)) {
						if (!conf.ignoreuser) {
							/* make sure maxbacklight is never <7 */
							if (backlight_restore < 7) conf.maxbacklight=7;
							else conf.maxbacklight=backlight_restore;
							if (verbose) printf("-> Detected user backlight change, setting maxbacklight to %d\n",conf.maxbacklight);
							backlight_prev=backlight_restore;
						} else {
							if (verbose) printf("-> Ignoring user backlight change, wants to set maxbacklight to %d\n",backlight_restore);
						}
						backlight=calculate_screen_backlight_value(light, conf.maxbacklight, conf.minbacklight);
					}
					backlight_restoreflag=1;
				}
				if (debug == 2 || debug == 3) printf ("-> set screen backlight: %d -> %d\n",backlight_prev,backlight);
				backlight_fading(backlight_prev, backlight, dbus_backend);
				usleep(1500);
				backlight=get_screen_backlight_value();
				backlight_prev=backlight;
			}
			if (!conf.manualmode) {
				tmp=get_screen_backlight_value();
				if (tmp!=backlight) {
					if (verbose) printf("-> Detected user backlight change, current backlight is set to %d\n", tmp);
					if (conf.ignoreuser) {
						if (debug == 2 || debug == 3) printf ("\n*** forcing backlight from %d to %d\n", tmp, backlight);
						backlight_fading(tmp, backlight, dbus_backend);
					}
				}
			}
		}

		if (verbose) printf("\n");

		usleep(conf.polltime*1000);
	}

	// we should never reach here.
	//if (conf.idleoff != 0) XCloseDisplay(display);
	//dbus_g_connection_unref(connection);
	exit(1);
}
